csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs

```diff
@@ -206,32 +209,108 @@ private void CanRunInferenceOnAModel(GraphOptimizationLevel graphOptimizationLev
+                    }
+                }
+
+                // Run inference with named inputs and named outputs
+                {
+                    // correct pre-allocated outputs
+                    var expectedOutputValues = new List<NamedOnnxValue>()
+                    {
+                        NamedOnnxValue.CreateFromTensor("softmaxout_1", new DenseTensor<float>(expectedOutputDimensions))
+                    };
+                    session.Run(container, expectedOutputValues);
+                    validateRunResultData(expectedOutputValues[0].AsTensor<float>());
+                }
+
+                // Run inference with pinned inputs and named outputs
+                using (var pinnedInputs = new DisposableList<PinnedOnnxValue>())
+                {
+                    var inputNames = container.Select(i => i.Name).ToArray();
+                    pinnedInputs.AddRange(container.Select(i => PinnedOnnxValue.CreateFromTensor(i.AsTensor<float>())));
+
+                    // expected inputs and outputs
+                    var expectedOutputValues = new List<NamedOnnxValue>()
+                    {
+                        NamedOnnxValue.CreateFromTensor("softmaxout_1", new DenseTensor<float>(expectedOutputDimensions))
+                    };
+                    session.Run(inputNames, pinnedInputs, expectedOutputValues);
+                    validateRunResultData(expectedOutputValues[0].AsTensor<float>());
+                }
+
+                // Run inference with named inputs and pinned outputs
+                {
+                    // correct pre-allocated outputs
+                    using (var pinnedOutputs = new DisposableList<PinnedOnnxValue>())
+                    {
+                        var outputTensor = new DenseTensor<float>(expectedOutputDimensions);
+                        pinnedOutputs.Add(PinnedOnnxValue.CreateFromTensor(outputTensor));
+                        session.Run(container, expectedOutputNames, pinnedOutputs);
+                        validateRunResultData(outputTensor);
+                    }
+                }
+
+                // Run inference with pinned inputs and pinned outputs
+                using (DisposableList<PinnedOnnxValue> pinnedInputs = new DisposableList<PinnedOnnxValue>(),
+                                                       pinnedOutputs = new DisposableList<PinnedOnnxValue>())
+                {
+                    var inputNames = container.Select(i => i.Name).ToArray();
+                    pinnedInputs.AddRange(container.Select(i => PinnedOnnxValue.CreateFromTensor(i.AsTensor<float>())));
+
+                    var outputTensor = new DenseTensor<float>(expectedOutputDimensions);
+                    pinnedOutputs.Add(PinnedOnnxValue.CreateFromTensor(outputTensor));
+
+                    session.Run(inputNames, pinnedInputs, expectedOutputNames, pinnedOutputs);
+                    validateRunResultData(outputTensor);
+                }
```

Comment:
For the 4 validateRunResultData() checks above, if any of checks fail,  the overall unit test will fail. Seems cleaner to create separate unit tests to pinpoint which check is failing.